### ISSUE TITLE
Create dynamic CTE (Common Table Expressions / WITH-clause) with model API

### ISSUE THREAD

**Issue by chilliger** [jOOQ/jOOQ#9895]
We have highly dynamic queries. While the [DSL API](https://www.jooq.org/doc/3.13/manual/sql-building/sql-statements/with-clause/) provides the ability to define a CTE table, there is no dynamic way to do so with the model API (at least I could not find one). The current way to do this dynamically is to hold a collection of CTE fields/names and the coressponding select query for the CTE table. It would be nice being able to add colums to the CTE table without the need to hold a separate collection. Somthing like:

```
cteTable = create.withQuery("aliasName");
c1 = cteTable.addColumn("c1");
cteTable.as(...);

selectQuery = create.selectQuery("t1");
selectQuery.setWith(cteTable);
selectQuery.addSelect(c1);
```

I also noticed that neither `create.with(...)` nor `DSL.name("t1").fields(...)` accept a collection. Maybe this is also something to improve.

Trying to implement a dynamic way with the current APIs. The problem here is that there is no way to add a `create.selectQuery` that references fields from the cts table with the DSL API `getCreator().with(...).select(...)`. So I am not able to combine the DSL API and the model but the cts table is currently only avaible in the DSL API. Bummer...

So basically right now I am in the need to build the complete WITH statement by myself with strings from the JOOQ objects. Everything else is inconvient.

**Comment by lukaseder** [comment]
Thank you very much for your suggestion. We're currently thinking about a much more thorough way to design the "model API" with a new query object model that is immutable:
https://github.com/jOOQ/jOOQ/issues/9163

With a complete re-design, CTEs will definitely be considered for better inclusion as well. The current model would make it difficult to expose CTEs in the `SelectQuery` API without significant refactorings, which aren't worth it considering that we're redesigning the model API anyway.

> I also noticed that neither `create.with(...)` nor `DSL.name("t1").fields(...)` accept a collection. Maybe this is also something to improve.

There's `DSLContext.with(CommonTableExpression...)`

> So basically right now I am in the need to build the complete WITH statement by myself with strings from the JOOQ objects. Everything else is inconvient.

The model API lends itself to an imperative style of constructing dynamic SQL. But that is not necessarily the only choice you have. A more functional style can be applied to the DSL API, to achieve much more elegant dynamic SQL construction.

Here are a few examples:

- https://blog.jooq.org/2018/02/27/map-reducing-a-set-of-values-into-a-dynamic-sql-union-query/
- https://blog.jooq.org/2017/01/16/a-functional-programming-approach-to-dynamic-sql-with-jooq/

Kotlin makes this even more elegant:

- https://blog.jooq.org/2017/06/09/using-kotlins-apply-function-for-dynamic-sql-with-jooq/

I think the investment in using jOOQ in a more FP than imperative way is a really rewarding one.

In any case, the model API should offer `WITH` support, and we'll definitely improve that when we r-design it. As such, I'm closing this issue as a duplicate of #9163

**Comment by chilliger** [comment]
@lukaseder sorry for responding late here.

> There's DSLContext.with(CommonTableExpression...)
This is what I meant with `create.with(...)`. No way to define that with a collection of field names.

I don't see how the linked resources and the DSL API can help in my case. I only want to provide a sophisticated `addSelect` to a model API `SelectQuery` that creates the `WITH` clause. Anything else is up to the client to add to that query e.g. conditions, order by clause, etc.

Anyway. I ended up with a little hack to add the  `WITH` clause to a `SelectQuery`:
```
CommonTableExpression<Record> withTable = withTableName
          .fields(defaultValueFields.toArray(new Name[0]))
          .as(defaultValueFromRow0OrXML);
      RenderContext ctx = getCreator().renderContext().declareCTE(true);
      String with = ctx.render(withTable);
      return getCreator().resultQuery("WITH " + with + " " + getCreator().render(select) + ";");
```

`getCreator().renderContext()` is deprecated. Found no other way to make a `CommonTableExpressionImpl` render correctly because it checks `ctx.declareCTE()` before rendering.
